### PR TITLE
Adding Standard Connections to the WDC Simulator

### PR DESCRIPTION
**Pull request by jimmy-jia**
This pull request adds support for Standard Connections to the WDC Simulator. It appears when you select the advanced UI. It creates tabs for each connection as well as the validation results and an interactive visualization of the standard connection.

### CODE REVIEW

**Comment by brandoconnor** [comment]
@jagreene can you have a look at this PR?

**Comment by jagreene** [comment]
I'll look at it as soon as I get back in the States on Monday

**Comment by jagreene** [comment]
We should figure out testing duties, now that people besides me are adding features. If you want to add the unit tests, I can add the end to end? Also, now I know that nyc doesn't check for coverage in files that don't have tests :P I'll fix that.

File: Simulator/actions/simulator_actions.js [hunk 1]
```diff
@@ -63,6 +68,7 @@ export function startConnector(phase) {
     dispatch(commitUrl());
     dispatch(closeSimulatorWindow());
     dispatch(setWindowAsExternal());
+    dispatch(resetStandardConnections());
```
Comment (jagreene) [inline Simulator/actions/simulator_actions.js]: (super nitpick) I might move the reset up to line 65 to be with the other reset

File: Simulator/components/JoinViz.jsx [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+          }
+        }
+      }
+    });
+  }
+}
+
+JoinViz.propTypes = {
+  tables: PropTypes.array.isRequired,
+  joins: PropTypes.array.isRequired,
```
Comment (jagreene) [inline Simulator/components/JoinViz.jsx]: I think you might be missing alias

File: Simulator/components/JoinViz.jsx [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    joins.forEach(join => {
+      const l = `[${join.left.tableAlias}].[${join.left.columnId}]`;
+      const r = `[${join.right.tableAlias}].[${join.right.columnId}]`;
+      data.edges.push({
+        from: tableMap[join.left.tableAlias],
+        to: tableMap[join.right.tableAlias],
+        joinValue: `[${l} ${join.joinType} joined with [${r}]]`,
+      });
+    });
+    const options = {
```
Comment (jagreene) [inline Simulator/components/JoinViz.jsx]: I might pull this out to Simulator/utils/consts.js and import it, just help keep this file leaner

File: Simulator/components/StandardConnections.jsx [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+      return true;
+    });
+    if (Object.keys(this.joinsVisitMap).length !== this.joins.length) {
+      this.errors.push('Not all joins visited!');
+    }
+  }
+
+  recurseTree(alias) {
+    this.tableVisitMap[alias] = true;
+    for (const i of this.joins) {
```
Comment (jagreene) [inline Simulator/components/StandardConnections.jsx]: I might change this to

`for (const join of this.joins) {`

for clarity

File: Simulator/components/App.jsx [hunk 5]
```diff
@@ -92,6 +97,16 @@ class App extends Component {
     const hasData = !!this.props.tables && Object.keys(this.props.tables).length > 0;
     const isAddressBarEmpty = (this.props.addressBarUrll === '');
 
+    const hasStandardConnections = this.props.standardConnections.length > 0;
+    if (hasStandardConnections) {
+      const connectionList = this.props.standardConnections;
+      standardConnectionsWindow = connectionList.map((standardConnection, idx) =>
```
Comment (jagreene) [inline Simulator/components/App.jsx]: (nitpick) I could be wrong but, this seems like a list of tabs, not a window so the name might be confusing

Comment (jimmy-jia) [inline Simulator/components/App.jsx]: Its a list of tabs that each have a standard connection window inside, naming it to something like standardConnectionsList / TabList would make more sense though.

**Comment by jimmy-jia** [comment]
Just committed some changes to address the comments from Austin (thanks for going through it!)

**Comment by jagreene** [comment]
looks like there are merge conflicts, would you mind rebasing off of the current dev, after that, and we figure out tests, I'd say it LGTM

**Comment by jimmy-jia** [comment]
Maybe we should wait on resolving the current conflicts until shortly before this is added to the dev branch? The exact timing on that hasn't been decided yet so we might have to rebase it a second time if other things are added before then. In the meantime we can decide on how to do the tests for this and get that ready.

**Comment by jagreene** [comment]
Yeah, good point. Somebody is going to have to rebase it before they can run the tests though because I merged in some test fixes to dev in the interim.

**Comment by jimmy-jia** [comment]
Ah I see, so we won't be able to test the test code without rebasing the new test fixes first?

**Comment by jagreene** [comment]
I mean you could but there will be extra failings and some of the failures were silent on GH

**Comment by jimmy-jia** [comment]
Well, in that case we should probably rebase it for the test case and again for the actual release if needed. Do you want to take some time tomorrow and talk about how we should test it?

**Comment by jagreene** [comment]
Sounds good!

File: Simulator/tests/unit/reducers.js [hunk 6]
```diff
@@ -166,6 +166,25 @@ describe("Reducer", function() {
     });
   });
 
+  describe("SET_STANDARD_CONNECTIONS", function() {
+    it("Should Set Standard Connections", function() {
+      const newValue = {
+        "id": {
+          standardConnections: [{ alias: "aliaa", tables: [], joins: [] }],
```
Comment (jagreene) [inline Simulator/tests/unit/reducers.js]: I'm guessing you meant alias not aliaa :P

File: Simulator/tests/unit/components.js [hunk 7]
```diff
@@ -338,4 +341,47 @@ describe("Components", function() {
...
+          alias="alias"
+          tables=[{id: "id1", alias: "alias1"},
+                  {id: "id2", alias: "alias2"}]
+          joins=[{left: {tableAlias: "alias1", columnId: "c1"},
+                  right: {tableAlias: "alias2", columnId: "c2"},
+                  joinType: "inner"}]
+        />
+      );
+      joinViz.should.be.ok();
+    })
```
Comment (jagreene) [inline Simulator/tests/unit/components.js]: I think these could use fleshing out. I would look at the tablePreview tests as an example of testing more complicated components with multiple interactions and member functions

File: Simulator/tests/unit/reducers.js (hunk 6 above)
Comment (jimmy-jia) [inline Simulator/tests/unit/reducers.js]: You would be right heh

**Comment by jagreene** [comment]
Probably want to get rid of the yep.png file :P

**Comment by brandoconnor** [comment]
100% coverage! Gold stars for @jimmy-jia and @jagreene !

**Pull request merged**